### DISCUSSION TITLE
cd: /usr/local/src/pivpn: No such file or directory

### DISCUSSION THREAD

**Discussion by mauromarano** [pivpn/pivpn#1409] [Support] [answered]
### Has your install failed? 
```
yes
```

#### What type of hardware are you running PiVPN at?
```
Raspberrypi (1 model B)

```
#### Output of `uname -a`
```
Linux octopi 5.4.79+ #1373 Mon Nov 23 13:18:15 GMT 2020 armv6l GNU/Linux
```

#### Output of `cat /etc/os-release`
```
PRETTY_NAME="Raspbian GNU/Linux 10 (buster)"
NAME="Raspbian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"
```

### If install failed Please provide the console output of  `curl -L https://install.pivpn.io | bash`
```
curl -L https://install.pivpn.io | bash
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed





100   162  100   162    0     0    201      0 --:--:-- --:--:-- --:--:--   201
100 82115  100 82115    0     0  45493      0  0:00:01  0:00:01 --:--:--  112k
:::
::: sudo will be used for the install.
::: Hostname length OK
::: Verifying free disk space...
:::
::: Package Cache update is needed, running apt-get update -y ...
 done!
:::
::: Checking apt-get for upgraded packages.... done!
:::
::: There are 9 updates available for your system!
::: We recommend you update your OS after installing PiVPN!
:::
:::    Checking for git... already installed!
:::    Checking for tar... already installed!
:::    Checking for curl... already installed!
:::    Checking for grep... already installed!
:::    Checking for dnsutils... already installed!
:::    Checking for whiptail... already installed!
:::    Checking for net-tools... already installed!
:::    Checking for bsdmainutils... already installed!
:::    Checking for dhcpcd5... already installed!
:::    Checking for iptables-persistent... already installed!
Reading package lists... Done
Building dependency tree
Reading state information... Done
0 upgraded, 0 newly installed, 0 to remove and 9 not upgraded.
::: Using User: pi
:::
::: Checking for existing base files...
:::    Checking /usr/local/src/pivpn is a repo...:::     Updating repo in /usr/local/src/pivpn...main: line 1007: cd: /usr/local/src/pivpn: No such file or directory
```

### Issue
Cannot complete the installation

### Console output of      `curl -L install.pivpn.io | bash`
```
curl -L install.pivpn.io | bash
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    17  100    17    0     0     40      0 --:--:-- --:--:-- --:--:--    40





100   162  100   162    0     0    143      0  0:00:01  0:00:01 --:--:--  5785
100 82115  100 82115    0     0  32063      0  0:00:02  0:00:02 --:--:--     0
:::
::: sudo will be used for the install.
::: Hostname length OK
::: Verifying free disk space...
:::
::: Package Cache update is needed, running apt-get update -y ...
 done!
:::
::: Checking apt-get for upgraded packages.... done!
:::
::: There are 9 updates available for your system!
::: We recommend you update your OS after installing PiVPN!
:::
:::    Checking for git... already installed!
:::    Checking for tar... already installed!
:::    Checking for curl... already installed!
:::    Checking for grep... already installed!
:::    Checking for dnsutils... already installed!
:::    Checking for whiptail... already installed!
:::    Checking for net-tools... already installed!
:::    Checking for bsdmainutils... already installed!
:::    Checking for dhcpcd5... already installed!
:::    Checking for iptables-persistent... already installed!
Reading package lists... Done
Building dependency tree
Reading state information... Done
0 upgraded, 0 newly installed, 0 to remove and 9 not upgraded.
::: Using User: pi
:::
::: Checking for existing base files...
:::    Checking /usr/local/src/pivpn is a repo... not found!
:::    Cloning https://github.com/pivpn/pivpn.git into /usr/local/src/pivpn...main: line 1029: cd: /usr/local/src/pivpn: No such file or directory
```

### Console output of      `pivpn add` or `pivpn add nopass`
```
pivpn add nopass
-bash: pivpn: command not found
```

### Console output of      `pivpn debug`

```
 pivpn debug
-bash: pivpn: command not found
```

**Comment by coolapso** [comment]
Moving to discussions: 

https://github.com/pivpn/pivpn/blob/master/README.md#general-guidelines

> The Github issues page is not intended for general support usage or feature requests

This error seems unusual, seems that the git clone failed for some reason

Are you doing this on a fresh install of raspbian? If not, Can you try this on a clean install?

**Comment by Peetah92** [comment]
I have the same issue. I tried to format the sd card and reinstalled OctoPrint on it, but the problem persists. Hope this could be resolved soon.

**Comment by coolapso** [comment, marked as answer]
The answer for this is here: https://docs.pivpn.io/faq/#what-about-octopi 

@Peetah92

**Reply by Peetah92** [reply]
Thank you very much! I was struggling with this for a month and now it works.

**Comment by megakoresh** [comment]
I am running this on Ubuntu 22.04 and running into same issue. There is no git wrapper in use.

```
megak@ubuntu-hel1-1:~$ sudo which git
/usr/bin/git
```

**Reply by coolapso** [reply]
Ubuntu 22.04 has not been tested yet, therefore its (for now) not a supported OS. try running `sudo apt update && sudo apt upgrade` before running the script.

**Reply by megakoresh** [reply]
I do it already. And the same issue happens on 20.04

**Reply by megakoresh** [reply]
@4s3ti I is  a bug. I [found](https://github.com/pivpn/pivpn/blob/f8cb945af15a1ca0cf063475c6e1557c6e8da06c/auto_install/install.sh#L1109) it. The `&` character puts the command into background and the cd instantly afterwards fails. It should be `&&`. Opened a PR and issue to fix.

**Reply by coolapso** [reply]
Thanks a lot. Merged into test branch :) I should push it to master later today.

**Reply by EPeterE** [reply]
Hi. Has this issue been resolved since I am having the same issue on Raspbian?

**Reply by megakoresh** [reply]
I have fixed it and the fix was merged to test, but it is still not deployed to master for whatever reason, so you need to use the test branch if you want a working installer

**Reply by ghost** [reply]
There are 7 instance of this bug
Some in `auto_install/install.sh` and some in `scripts/uninstall.sh`

**Reply by ghost** [reply]
See my commits for details
https://github.com/pivpn/pivpn/pull/1572/commits/47548c7ca22b348eae0564dc37095df22854ea48
https://github.com/pivpn/pivpn/pull/1567/commits/fb780830da1f267c3fe0d5c70874ae153f3e23da

**Reply by megakoresh** [reply]
You know I am starting to think this is not entirely a bug. I think this is actually intended, and it's just the spinner function which is broken. Or maybe it must be invoked on next line because the spinner is supposed to capture the pid of the previous background task and NOT return until it's finished. But I think it doesn't actually do that and returns instantly and thus we have the bug.

I think the spinner is not necessary it would probably be better to just remove it so that various subtle shell and distro differences don't break the compatibility of the installer. What do you think @4s3ti @giulio-coa

**Reply by ghost** [reply]
I think you are right
The PR tests associated with the two commits have been running for over 10 minutes ...

**Reply by ghost** [reply]
I confirm that by adding a semicolon, the bug should be fixed
`... & spinner ...` -> `... &; spinner ...`

**Reply by EPeterE** [reply]
Hi
Using the test/development branch "curl -L https://test.pivpn.io | TESTING= bash" gives the same error
::: Checking for existing base files...
:::    Checking /usr/local/src/pivpn is a repo... not found!
:::    Cloning https://github.com/pivpn/pivpn.git into /usr/local/src/pivpn ...main: line 1137: cd: /usr/local/src/pivpn: No such file or directory

**Reply by ghost** [reply]
Now the test branch should be fixed
